ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,314 @@
+import rclpy
+from rclpy.node import Node
+from sensor_msgs.msg import Image
+import os
+from threading import Lock
+from typing import Tuple
+from rclpy.parameter import Parameter
+from rcl_interfaces.msg import ParameterDescriptor, ParameterType
+
+
+from ada_feeding_perception.helpers import (
+    download_checkpoint,
+)
+
+
+class FaceDetectionNode(Node):
+    def __init__(
+        self,
+    ):
+        """
+        Initializes the FaceDetection node. This node exposes a ToggleFaceDetection
+        service that can be used to toggle the face detection on or off and
+        publishes information about detected faces to the /face_detection
+        topic when face detection is on.
+        """
+        super().__init__("face_detection")
+
+        # Read the parameters
+        # NOTE: These parameters are only read once. Any changes after the node
+        # is initialized will not be reflected.
+        (
+            face_model_name,
+            face_model_base_url,
+            landmark_model_name,
+            landmark_model_base_url,
+            model_dir,
+        ) = self.read_params()
+        self.face_model_name = face_model_name.value
+        self.landmark_model_name = landmark_model_name.value
+
+        # Download the checkpoints if they don't exist
+        self.face_model_path = os.path.join(model_dir.value, self.face_model_name)
+        if not os.path.isfile(self.face_model_path):
+            self.get_logger().info(
+                "Face detection model checkpoint does not exist. Downloading..."
+            )
+            download_checkpoint(
+                self.face_model_name, model_dir.value, face_model_base_url.value
+            )
+            self.get_logger().info(
+                "Model checkpoint downloaded %s."
+                % os.path.join(model_dir.value, self.face_model_name)
+            )
+        self.landmark_model_path = os.path.join(
+            model_dir.value, self.landmark_model_name
+        )
+        if not os.path.isfile(self.landmark_model_path):
+            self.get_logger().info(
+                "Facial landmark model checkpoint does not exist. Downloading..."
+            )
+            download_checkpoint(
+                self.landmark_model_name, model_dir.value, landmark_model_base_url.value
+            )
+            self.get_logger().info(
+                "Model checkpoint downloaded %s."
+                % os.path.join(model_dir.value, self.landmark_model_name)
+            )
+
+        # Convert between ROS and CV images
+        self.bridge = CvBridge()
+
+        # Keeps track of whether face detection is on or not
+        self.is_on = False
+        self.is_on_lock = Lock()
+
+        # Keeps track of the detected mouth center
+        self.img_mouth_center = []
+        self.is_face_detected = False
+
+        # Create the service
+        self.srv = self.create_service(
+            ToggleFaceDetection,
+            "ToggleFaceDetection",
```

Comment:
```suggestion
            "toggle_face_detection",
```
[ROS naming conventions](http://wiki.ros.org/ROS/Patterns/Conventions) have topics and services be in snake_case. (I know this was a carryover from the dummy node -- that was my mistake)